gatsby/content/blog/2021/12/2021-12-10-sydent-typing-2.mdx

```diff
@@ -0,0 +1,384 @@
+
+I think it would have been more efficient to use `stubgen` to generate stubs and patch them up, rather than writing them. That would have helped avoid a few cases I encountered where stubbing one function would cause additional typechecking failures (because mypy was no longer examining the twisted source for that file). It would also have meant that I could just faithfully stub Twisted as it is; in practice, I would sometimes hesitate to stub to avoid having to cover another module. `sydent.http` was the most painful part of the source tree for this: that's where we make the most use of Twisted.
+
+#### [`defer.inlineCallbacks`](https://twistedmatrix.com/documents/current/api/twisted.internet.defer.html#inlineCallbacks)
+
+This is a decorator which allows us to write code in the style of `async`/`await` without actually using that syntax. (Twisted predates `asyncio` introduced in Python 3.4; it was originally released in 2002, back when Python had released version 2.2.) We only use it in [one place in Sydent](https://github.com/matrix-org/sydent/blob/e4b4dbbdf25255d19effceae7061ad9b3eefc374/sydent/http/matrixfederationagent.py#L125-L132) nowadays, but I've seen used across Synapse too. I mention it here because it was a bit fiddly to annotate. Here's its use in Sydent:
+
+```python
+    @defer.inlineCallbacks
+    def request(
+        self,
+        method: bytes,
+        uri: bytes,
+        headers: Optional["Headers"] = None,
+        bodyProducer: Optional["IBodyProducer"] = None,
+    ) -> Generator["defer.Deferred[Any]", Any, IResponse]:
+```
+
+Here, `request` is a [generator function](https://docs.python.org/3/glossary.html#term-generator) because its body uses the `yield` keyword. Yielding allows the function to relinquish control back to twisted's reactor, only for its execution to be resumed asynchronously in the future. The [`Generator`](https://docs.python.org/3/library/typing.html?highlight=typing%20generator#typing.Generator) type takes three parameters: a `YieldType`, a `SendType`, and a `ReturnType`. In this example:
+
+- We yield a `Deferred[Any]`.
+- That will later be resolved by twisted to an `Any` value. Twisted will send that value to `request`, and the execution continues.
+- This repeats, until we eventually return an `IResponse`.
+
+(I'm guessing `x = await y` is really some kind of sugar around this mechanism: first `yield y`, then the machinery running your coroutine `c` will call `c.send(x)` to resume execution.)
```

Comment:
This seems like it's referencing a code snippet that I haven't seen yet?

Reply:
I've tried to improve this in 54ac7583. What do you think @callahad ?